### PR TITLE
[WIP] Update entities.xml

### PR DESCRIPTION
**Pull request by gashcrumb**

### CODE REVIEW

**Comment by gashcrumb** [comment]
@kahboom FYI worth a review...

**Comment by gashcrumb** [comment]
@kahboom FYI, last commit consolidates the join tables 'IntegrationTemplatesConnections' and 'IntegrationTemplatesSteps' into one table, same for the ones that relate Integrations to Connections and Steps.  I replaced `step_position` with `previous_step_id` and `next_step_id`, so each row tracks it's position in the overall route.  Also added a `type` enum column so it's easy to determine if a given row is a Connection or a Step.  Think this will let us handle potentially more complicated routes.

Now to look at adding `Projects`

**Comment by kahboom** [comment]
@gashcrumb - Thanks! Are you a huge proponent of snake case or would you be willing to use camelcase like we do when defining Sequelize model attributes?

File: docs/entities.md [hunk 1]
```diff
@@ -4,49 +4,86 @@ We are in the process of creating a diagram (see below) for the Entity Relations
...
 - IntegrationTemplate
-    - has many Connections
-    - has many Steps
+    - has many Connections through IntegrationTemplateConnectionStep
+    - has many Steps through IntegrationTemplateConnectionStep
     - has many Tags
+    - belongs to an Organization
+
+- IntegrationTemplateConnectionStep
+  - has a previous step ID
```
Comment (kahboom) [inline docs/entities.md]: I don't think we should be tracking attributes through here tbh.

File: docs/entities.md [hunk 2]
```diff
@@ -86,3 +123,17 @@ Again, this list may not be up-to-date, or may be under development. Please see
 ## Visual entity graph
 
 ![Visual entity graph picture](./entities.png)
+
+You can edit the diagram of the above relationships using [this online tool](http://ondras.zarovi.cz/sql/demo/?keyword=default)
```
Comment (kahboom) [inline docs/entities.md]: Cool, I like this!! I might be moving all of these docs to the main RH iPaaS website later on like we discussed. But I'll definitely have to reference it from the repository.

File: docs/entities.md (hunk 1 above)
Comment (gashcrumb) [inline docs/entities.md]: yeah, will take those out

**Comment by gashcrumb** [comment]
Yeah, lemme change that stuff to camel case

File: docs/entities.md (hunk 1 above)
Comment (gashcrumb) [inline docs/entities.md]: Oh I see, we're not tracking any attributes in this file, that's cool.

**Comment by kahboom** [comment]
@gashcrumb - Sorry, merged it in without realizing that you hadn't added Projects yet. I can go ahead and do that if you'd like?

**Comment by gashcrumb** [comment]
Oh, no projects is in there too, should be in the diagram and entities etc.

**Comment by kahboom** [comment]
Ah gotcha, I meant the model itself. I'll add it now though!

Regards,

Rachel Yordán
Senior Software Engineer
Red Hat, Inc. | Middleware
ryordan@redhat.com

**Comment by gashcrumb** [comment]
Oh shoot yeah I haven't changed any of that yet, wanted to be sure the
diagram and entities were good to go first.  If you have a chance though go
for it, thanks!

**Comment by kahboom** [comment]
@gashcrumb Oh wait but you are associating them with Users, not Organizations? Sorry I edited my comment above so I don't think you got notified of it via email, let me copy and paste:

>Ah gotcha, I meant the model itself. I'll add it now though! The only thing is that I don't know if projects should belong to organizations or users.. hrmmm.. I think it should belong to the organization but should also be associated with a user, like a user can have many projects they've created, but ultimately it belongs to the organization, not the user (in case they ever leave, for instance).

What do you think?

**Comment by gashcrumb** [comment]
That makes sense, yeah, was wondering if/how to associate projects and
organizations, would make sense to have that association for sure.

**Comment by kahboom** [comment]
@gashcrumb - Okay cool, then I'll make it so that each Project belongs to an Organization, but can have many Users (that can be subsequently kicked out if they leave the Organization, etc.).

**Comment by gashcrumb** [comment]
Sounds good!  I can update the diagram tomorrow morning...

**Comment by kahboom** [comment]
@gashcrumb - So, I'm updating all of the models, and some we still hadn't included like Component, ComponentGroup, IntegrationPattern, etc. The one I'm most confused about is the triple-JOIN one - `IntegrationTemplatesConnectionsSteps` rofl...  also `IntegrationsConnectionsSteps`. So, maybe we should just call them `IntegrationTemplatesConnections` and `IntegrationsConnections`, we can still include the `StepId` as an attribute if anything. It's just that it takes me a few seconds to realize the association, although having all models in the table name could make it more obvious. Not sure tbh.. either way we'd need these models carved out by tomorrow morning so I can continue with the UI and start re-doing the dummy data using our new models.

**Comment by gashcrumb** [comment]
Yeah, as horridly long as the table name is, it's naming is consistent with
all the others.  And you're right, we've still got to update/rename/add
some stuff according to the diagram, but wanted to be sure that was settled
before committing it to code, you know?

I like having steps and connections tracked from the one join table, that
means 1 query to fetch all the items for an Integration or
IntegrationTemplate, and each row that's fetched tells you that it's either
a connection or step so you can then easily fetch those from the other two
tables, and you also have the placement of that row in relation to it's
neighbors.

**Comment by kahboom** [comment]
@gashcrumb - Okay, all good points. I'll leave it like `IntegrationTemplatesConnectionsSteps`. Like we discussed, it's usually better to have more tables when in doubt lol. I'm adding this as a 'through' string and we'll see if Sequelize accepts it for all three associations or if I have to define some type of foreign key constraint. Should be fine though! 👍

**Comment by gashcrumb** [comment]
Ah yeah, that's the crux really, if sequelize will let us do it, otherwise
we might have to maintain that association in the implementation rather
than as part of the DB schema, you know?  :-)

**Pull request merged**